File: windows/AirBridge.Core/InMemoryDeviceRegistry.cs
using AirBridge.Core.Interfaces;
using AirBridge.Core.Models;
using System.Collections.Concurrent;

namespace AirBridge.Core;

/// <summary>
/// Thread-safe in-memory implementation of <see cref="IDeviceRegistry"/>.
/// Devices are not persisted across restarts; paired state is stored in KeyStore.
/// </summary>
public sealed class InMemoryDeviceRegistry : IDeviceRegistry
{
    private readonly ConcurrentDictionary<string, DeviceInfo> _devices = new();

    /// <inheritdoc/>
    public event EventHandler<DeviceInfo>? DeviceChanged;

    /// <inheritdoc/>
    public IReadOnlyList<DeviceInfo> GetAllDevices()
        => _devices.Values.ToList();

    /// <inheritdoc/>
    public IReadOnlyList<DeviceInfo> GetPairedDevices()
        => _devices.Values.Where(d => d.IsPaired).ToList();

    /// <inheritdoc/>
    public void AddOrUpdate(DeviceInfo device)
    {
        ArgumentNullException.ThrowIfNull(device);
        _devices[device.DeviceId] = device;
        DeviceChanged?.Invoke(this, device);
    }

    /// <inheritdoc/>
    public void Remove(string deviceId)
    {
        if (_devices.TryRemove(deviceId, out var removed))
            DeviceChanged?.Invoke(this, removed);
    }

    /// <inheritdoc/>
    public bool IsPaired(string deviceId)
        => _devices.TryGetValue(deviceId, out var d) && d.IsPaired;
}

File: android/app/src/main/java/com/airbridge/app/core/registry/InMemoryDeviceRegistry.kt
package com.airbridge.app.core.registry

import com.airbridge.app.core.interfaces.IDeviceRegistry
import com.airbridge.app.core.models.DeviceInfo
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Thread-safe in-memory implementation of [IDeviceRegistry].
 *
 * Devices are stored in a [ConcurrentHashMap] keyed by [DeviceInfo.deviceId].
 * [devicesFlow] emits a fresh snapshot whenever the map changes.
 */
@Singleton
class InMemoryDeviceRegistry @Inject constructor() : IDeviceRegistry {

    private val _map = ConcurrentHashMap<String, DeviceInfo>()
    private val _stateFlow = MutableStateFlow<Map<String, DeviceInfo>>(emptyMap())

    /** Reactive stream of all known devices — emits on every add/update/remove. */
    override val devicesFlow: Flow<List<DeviceInfo>>
        get() = _stateFlow.asStateFlow().map { it.values.toList() }

    override fun getAllDevices(): List<DeviceInfo> = _map.values.toList()

    override fun getPairedDevices(): List<DeviceInfo> =
        _map.values.filter { it.isPaired }

    override fun addOrUpdate(device: DeviceInfo) {
        _map[device.deviceId] = device
        _stateFlow.value = _map.toMap()
    }

    override fun remove(deviceId: String) {
        _map.remove(deviceId)
        _stateFlow.value = _map.toMap()
    }

    override fun isPaired(deviceId: String): Boolean =
        _map[deviceId]?.isPaired == true
}
